doc/csharp/csharp-null-ref-types-proposal.md

```diff
@@ -0,0 +1,197 @@
+
+The generated code is not in a nullable context. The `// <auto-generated>` 
+comment in the code disables it. No nullable checks are made and no compiler 
+warnings are generated for this code, even if the code is compiled in a project 
+with nullable checks enabled.
+
+### 2. .NET Framework projects default to using C# 7.3
+
+*.NET Framework* projects default to using compiler version C# 7.3 which does
+not support nullable reference type features.
+
+*Visual Studio 2022* does not let you change to a later compiler version for
+*.NET Framework* projects:
+- In *Advanced Build Settings | Language version* the drop down to select the 
+language version is greyed out and says “*Automatically selected based on 
+framework version*”
+
+See https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version 
+
+Note: It is possible to edit the project file and explicitly set the language 
+version to a later version, but other language features that require runtime 
+changes in the CLR will not work. C# 7.3 is the latest language version that is 
+fully supported with *.NET Framework* projects.
+
+### 3. Building gRPC projects with nullable enabled works today
+
+There does not appear to be a problem building *.NET Core* projects that have
+nullable enabled and are using the generated code. All the `grpc/dotnet-grpc`
+examples already have nullable context enabled and they build without errors 
+or warnings.
+
+### 4. Breaking backwards compatibility with gRPC service classes
+
+The generated code for gRPC services could have attributes added to specify that 
+a non-null message or stream must be passed to a service, and a non-null value 
+is returned.
+
+However this can break existing user code. Adding these attributes to the base 
+class may cause errors (not warnings) in user code that has already been derived 
+from the base class:
+
+> *error CS8765: Nullability of type of parameter 'request' doesn't match 
+overridden member (possibly because of nullability attributes).*
+
+Note: the advantages of adding these attributes to the service methods is 
+questionable as the only caller of these methods is the server hosting the 
+services, not user code. Therefore this may not be an issue if it is not needed 
+to be done.
+
+### 5. Keeping the generated code simple and consistent
+
+It would be possible to generate different code for nullable and non-nullable contexts
+depending on some configuration, or to generate conditional code containing `#if` 
+everywhere.
+
+Generating different code depending on configuration could lead to confusion or errors
+if code generated in one project with nullable enabled was consumed in another project 
+that did not have nullable enabled. It would also have an impact on the level of 
+testing that would need to be done.
+
+Generating conditional code would produce messy and potentially convoluted code, again
+having an impact on testing.
```

Comment:
Yes, but I think we should figure out how big the difference will be between nullable enabled generated code and nullable disabled generated code.

I don't think there are many places in generated source that would be annotated as nullable. I took a quick look through some generated source code:

* Message fields can be null.
* `Equals(obj)` and `Equals(T)` would have nullable parameters.
* `MergeFrom(T)` would have a nullable parameter.
* `BindService` would have a nullable parameter (we could skip annotating this one since the only caller that would pass null is `Grpc.AspNetCore.Server`).

I doubt that is everything, but I don't think there is a lot more.

Reply:
There are a few other places in the generated code that break when nullable enabled, e.g.
```
public virtual global::Attrtest.Person FindPerson(
   global::Attrtest.SearchCriteria request,
   grpc::Metadata headers = null,
   global::System.DateTime? deadline = null,
   global::System.Threading.CancellationToken cancellationToken = default(global::System.Threading.CancellationToken)
   )
      {
        return FindPerson(request, new grpc::CallOptions(headers, deadline, cancellationToken));
      }
```
the `grpc::Metadata headers = null` is one such place above that needs changing to make it compile without warnings.

It just gets messy.

Reply:
I updated my list with `headers`.